src/librustc_typeck/check/coercion.rs

```diff
@@ -907,7 +917,12 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         let (a_sig, b_sig) = {
             let is_capturing_closure = |ty| {
                 if let &ty::Closure(_, substs) = ty {
-                    substs.as_closure().upvar_tys().next().is_some()
+                    let ty = self.infcx.shallow_resolve(substs.as_closure().upvar_tuple_ty());
```

Comment:
Check upvars_mentioned here. We can answer the question if the closure is capturing or not without using substs.